aiomysql/connection.py

```diff
@@ -767,35 +785,174 @@ def insert_id(self):
-            # clear-text-authentication.html
-            data = self._password.encode('latin1') + b'\0'
+        # These auth plugins do their own packet handling
+        if plugin_name == b"caching_sha2_password":
+            await self.caching_sha2_password_auth(auth_packet)
+            self._auth_plugin_used = plugin_name.decode()
+        elif plugin_name == b"sha256_password":
+            await self.sha256_password_auth(auth_packet)
+            self._auth_plugin_used = plugin_name.decode()
         else:
+
+            if plugin_name == b"mysql_native_password":
+                # https://dev.mysql.com/doc/internals/en/
+                # secure-password-authentication.html#packet-Authentication::
+                # Native41
+                data = _auth.scramble_native_password(
+                    self._password.encode('latin1'),
+                    auth_packet.read_all())
+            elif plugin_name == b"mysql_old_password":
+                # https://dev.mysql.com/doc/internals/en/
+                # old-password-authentication.html
+                data = _auth.scramble_old_password(
+                    self._password.encode('latin1'),
+                    auth_packet.read_all()
+                ) + b'\0'
+            elif plugin_name == b"mysql_clear_password":
+                # https://dev.mysql.com/doc/internals/en/
+                # clear-text-authentication.html
+                data = self._password.encode('latin1') + b'\0'
+            else:
+                raise OperationalError(
+                    2059, "Authentication plugin '{0}'"
+                          " not configured".format(plugin_name)
+                )
+
+            self.write_packet(data)
+            pkt = await self._read_packet()
+            pkt.check_error()
+
+            self._auth_plugin_used = plugin_name.decode()
+
+            return pkt
+
+    async def caching_sha2_password_auth(self, pkt):
+        # No password fast path
+        if not self._password:
+            self.write_packet(b'')
+            pkt = await self._read_packet()
+            pkt.check_error()
+            return pkt
+
+        if pkt.is_auth_switch_request():
+            # Try from fast auth
+            logger.debug("caching sha2: Trying fast path")
+            self.salt = pkt.read_all()
+            scrambled = _auth.scramble_caching_sha2(
+                self._password.encode('latin1'), self.salt
+            )
+
+            self.write_packet(scrambled)
+            pkt = await self._read_packet()
+            pkt.check_error()
+
+        # else: fast auth is tried in initial handshake
+
+        if not pkt.is_extra_auth_data():
             raise OperationalError(
-                2059, "Authentication plugin '%s' not configured" % plugin_name
+                "caching sha2: Unknown packet "
+                "for fast auth: {0}".format(pkt._data[:1])
             )
 
+        # magic numbers:
+        # 2 - request public key
```

Comment:
Good idea to make those numbers into enum, but I am fine merging PR as it is.

Reply:
I'd agree, but leaving them as they are does give us the benefit that it looks near identical to the equivalent pymysql function.